### PR TITLE
PS-3967: Use MAP_POPULATE when NUMA interleave is turned on. (5.7)

### PR DESCRIPTION
**Pull request by dutow**
Compared to the NUMA implementation in MySQL, this version forces allocation using MAP_POPULATE.

(cherry picked from commit 1170aef6fa012e44398e8dfa3096b440a3a77fda)

### CODE REVIEW

File: storage/innobase/buf/buf0buf.cc [hunk 1]
```diff
@@ -1484,7 +1484,8 @@ buf_chunk_init(
 /*===========*/
 	buf_pool_t*	buf_pool,	/*!< in: buffer pool instance */
 	buf_chunk_t*	chunk,		/*!< out: chunk of buffers */
-	ulint		mem_size)	/*!< in: requested size in bytes */
+	ulint		mem_size,	/*!< in: requested size in bytes */
+	ibool		populate)	/*!< in: virtual page preallocation */
```
Comment (percona-ysorokin) [inline storage/innobase/buf/buf0buf.cc]: The same `ibool`->`bool` comments as in 5.6

File: storage/innobase/os/os0proc.cc [hunk 2]
```diff
@@ -53,6 +66,24 @@ my_bool	os_use_large_pages;
...
 
+
+/****************************************************************//**
+Retrieve and compare operating system release.
+@return	TRUE if the OS release is equal to, or later than release. */
+ibool
+os_compare_release(
+/*===============*/
+	const char*	release		/*!< in: OS release */
+	__attribute__((unused)))
```
Comment (percona-ysorokin) [inline storage/innobase/os/os0proc.cc]: MY_ATTRIBUTE((unused))

File: storage/innobase/os/os0proc.cc [hunk 3]
```diff
@@ -146,17 +178,37 @@ os_mem_alloc_large(
...
 	}
 #endif
+
+#if OS_MAP_ANON && OS_MAP_POPULATE
+	/* MAP_POPULATE is only supported for private mappings
+	since Linux 2.6.23. */
+	populate = populate && !os_compare_release("2.6.23");
+
+	if (populate) {
+		fprintf(stderr, "InnoDB: Warning: mmap(MAP_POPULATE) "
```
Comment (percona-ysorokin) [inline storage/innobase/os/os0proc.cc]: new Logging mechanism in 5.7
`ib::fatal() << ` / `ib::error() << ` / `ib::warn() << `

File: storage/innobase/srv/srv0srv.cc [hunk 4]
```diff
@@ -269,6 +269,8 @@ with mutex_enter(), which will wait until it gets the mutex. */
 
 /** Requested size in bytes */
 ulint	srv_buf_pool_size	= ULINT_MAX;
+/* force virtual page preallocation (prefault) */
+my_bool	srv_buf_pool_populate	= FALSE;
```
Comment (percona-ysorokin) [inline storage/innobase/srv/srv0srv.cc]: not needed any more

File: storage/innobase/include/ut0new.h [hunk 5]
```diff
@@ -605,7 +605,8 @@ class ut_allocator {
 	pointer
 	allocate_large(
 		size_type	n_elements,
-		ut_new_pfx_t*	pfx)
+		ut_new_pfx_t*	pfx,
+		bool           populate)
```
Comment (percona-ysorokin) [inline storage/innobase/include/ut0new.h]: formatting

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

File: storage/innobase/buf/buf0buf.cc [hunk 6]
```diff
@@ -2864,7 +2868,8 @@ buf_pool_resize()
 			while (chunk < echunk) {
 				ulong	unit = srv_buf_pool_chunk_unit;
 
-				if (!buf_chunk_init(buf_pool, chunk, unit)) {
+				if (!buf_chunk_init(buf_pool, chunk, unit,
+				                    (bool) srv_numa_interleave)) {
```
Comment (percona-ysorokin) [inline storage/innobase/buf/buf0buf.cc]: static_cast<bool>(srv_numa_interleave)

File: storage/innobase/srv/srv0srv.cc [hunk 7]
```diff
@@ -269,6 +269,8 @@ with mutex_enter(), which will wait until it gets the mutex. */
 
 /** Requested size in bytes */
 ulint	srv_buf_pool_size	= ULINT_MAX;
+/* force virtual page preallocation (prefault) */
+my_bool	srv_buf_pool_populate	= false;
```
Comment (percona-ysorokin) [inline storage/innobase/srv/srv0srv.cc]: Not removed

**Review by percona-ysorokin — CHANGES_REQUESTED** [review]

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Comment by percona-ysorokin** [comment]
Jenkins build link for custom Percona Server 5.7.21-21 with NUMA patches
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1797/

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
s/MAP_PREPOPULATE/MAP_POPULATE in the commit message

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by dutow** [comment]
Jenkins: https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1806/

**Pull request merged**